### ISSUE TITLE
Improve Data Fetching in Unity

### ISSUE THREAD

**Issue by david-rhodes** [mapbox/mapbox-sdk-cs#11]
# TL;DR
Our current Unity implementation for making web requests is quite inefficient. Let's find ways to improve it.

# Possible Inefficiencies
- Web requests are using coroutines. Each coroutine allocates memory on the heap (every tick). Requesting hundreds of tiles will lead to "massive" GC spikes.
- Requests are not threaded.
- Each low level object that wants to make a request needs a `FileSource`. This is often duplicated between objects. Unknown what the correct solution is, but probably depends on our threading options.
- Currently, response errors are handled as strings. String compares are expensive and prone to error.

# Other Considerations
- Coroutines do not operate (sans hacks) at edit time. This makes it difficult to perform requests outside of "play" mode.
- I expect developers to hit their rate limits quite easily. It would be nice if we offered a convenience layer to automatically handle the distribution of requests or re-queuing of requests on rate limit exceeded errors. Of course a lot of developers will implement their own version of this to suit their needs, but ours may serve as a solid starting point.
- Web requests should have timeouts (I have, on occasion, observed requests never providing a response).

# Possible Solutions
- Use something like [More Effective Coroutines](https://www.assetstore.unity3d.com/en/#!/content/54975) (note: unsure if this will also solve the editor coroutine issues, but it could since you don't need to inherit from MonoBehaviour).
- Use [C# Threads](https://msdn.microsoft.com/en-us/library/system.threading.thread(v=vs.110).aspx).
- Use [System.Web](https://msdn.microsoft.com/en-us/library/gg145018(v=vs.110).aspx) for web requests (possibly less overhead).
- @BergWerkGIS mentioned using  [BruTile ](https://github.com/BruTile/BruTile) as a possible library for managing batch requests.

# Next Steps
It would be great to have a C# expert (@BergWerkGIS, I'm looking at you) experiment in this area.

/cc @mapbox/games
